app/models/piece.rb

```diff
@@ -1,8 +1,147 @@
-class Piece < ApplicationRecord
-  belongs_to :user
-  belongs_to :game
-
-  def self.types
-    %w[King Queen Rook Bishop Knight Pawn]
-  end
-end
+class Piece < ApplicationRecord
+  belongs_to :user
+  belongs_to :game
+
+  def self.types
+    %w[King Queen Rook Bishop Knight Pawn]
+  end
+
+  def obstructed?(destination_x, destination_y)
+    raise 'Error: Invalid Input' if destination_y > 7 || destination_x > 7 || destination_y < 0 || destination_x < 0
+    invalid_move_tests =
+      [
+        horizontal_move?(destination_y),
+        vertical_move?(destination_x),
+        diagonal_up_and_right_move?(destination_x, destination_y),
+        diagonal_down_and_left_move?(destination_x, destination_y),
+        diagonal_up_and_left_move?(destination_x, destination_y),
+        diagonal_down_and_right_move?(destination_x, destination_y)
+      ]
+    raise 'Error: Invalid Input' unless invalid_move_tests.include?(true)
+    destination_tests = []
+    invalid_move_tests.each_with_index do |test, index|
+      if test == true && index == 0
+        destination_tests << horizontal_obstruction_left?(destination_x)
+        destination_tests << horizontal_obstruction_right?(destination_x)
+      elsif test == true && index == 1
+        destination_tests << vertical_obstruction_up?(destination_y)
+        destination_tests << vertical_obstruction_down?(destination_y)
+      elsif test == true && index == 2
+        destination_tests << diagonal_obstruction_up_right?(destination_y)
+      elsif test == true && index == 3
+        destination_tests << diagonal_obstruction_down_left?(destination_y)
+      elsif test == true && index == 4
+        destination_tests << diagonal_obstruction_up_left?(destination_y)
+      elsif test == true && index == 5
+        destination_tests << diagonal_obstruction_down_right?(destination_y)
+      end
+    end

```

Comment:
This is getting really close. You can simplify this code more to make it easier to follow.

Try a pattern like this

```Ruby
def obstructed?(destination_x, destination_y)
  raise 'Error: Invalid Input' if destination_y > 7 || destination_x > 7 || destination_y < 0 || destination_x < 0

  # if is a certain type of move, check if that direction is obstructed.

  if diagonal_down_and_right_move?(destination_x, destination_y)
    return diagonal_obstruction_down_right?(destination_y)
  end

  #... test other directions

  if horizontal_move?(destination_y)
    return horizontal_obstruction_left?(destination_x) && horizontal_obstruction_right?(destination_x)
  end

  raise 'Error: Invalid Input'
end
```

When thinking about a method. Try to make it read like a paragraph. Another a developer should be able to come along and read top-to-bottom and mostly understand what it's doing.